src/read_shard.rs

```diff
@@ -1,3 +1,279 @@
+        unimplemented!()
+    }
+
+    fn handle_get_client_shard_info_request(
+        &self,
+        _req: &GetClientShardInfoRequest,
+    ) -> GetClientShardInfoResponse {
+        unimplemented!()
+    }
+
+    fn handle_query_version_request(&self, _req: &QueryVersionRequest) -> QueryVersionResponse {
+        unimplemented!()
+    }
+
+    fn handle_get_version_request(&self, _req: &GetVersionRequest) -> GetVersionResponse {
+        unimplemented!()
+    }
+
+    fn handle_write_request(&self, _req: &WriteRequest) -> WriteResponse {
+        unimplemented!()
+    }
+
+    fn handle_get_shared_peers_request(
+        &self,
+        _req: &GetSharedPeersRequest,
+    ) -> GetSharedPeersResponse {
+        unimplemented!()
+    }
+
+    fn handle_get_client_shard_info_response(&self, _res: &GetClientShardInfoResponse) {
+        unimplemented!()
+    }
+
+    fn handle_read_response(&self, _res: &ReadResponse) {
+        unimplemented!()
+    }
+
+    fn handle_write_response(&self, _res: &WriteResponse) {
+        unimplemented!()
+    }
+}
+
+impl ReadShard {
+    pub fn new(reader_ip_port: Arc<Mutex<([u8; 16], u16)>>) -> ReadShard {
+        ReadShard {
+            _reader_ip_port: reader_ip_port,
+            writer_id: Arc::new(Mutex::new(0)),
+            writer_ip_port: Arc::new(Mutex::new(([0; 16], 0))),
+            peers: Arc::new(Mutex::new(Vec::new())),
+            requested_version: Arc::new(Mutex::new(0)),
+            current_version: Arc::new(Mutex::new(0)),
+            history: Arc::new(Mutex::new(Vec::new())),
+            data: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+pub fn socket_addr_to_string(addr: ([u8; 16], u16)) -> String {
+    format!(
+        "{}:{}",
+        addr.0
+            .iter()
+            .map(|x| x.to_string())
+            .collect::<Vec<String>>()
+            .join("."),
+        addr.1
+    )
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let reader_ip_port = ([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1], 8084);
+
+    let info_router = ReadShard::new(Arc::new(Mutex::new(reader_ip_port)));
+    let info_router_1 = info_router.clone();
+    let info_server = RouterBuilder::new(info_router_1, None);
+
+    let client1 = info_server.get_router_client();
+    tokio::spawn(async move {
+        let mut interval = tokio::time::interval(tokio::time::Duration::from_secs(3));
+        loop {
+            interval.tick().await;
+
+            let announce_request = AnnounceShardRequest {
+                shard_type: 1,
+                ip: reader_ip_port.0,
+                port: reader_ip_port.1,
+            };
+
+            let _ = client1
+                .queue_request::<AnnounceShardRequest>(
+                    announce_request,
+                    socket_addr_to_string(MAIN_INSTANCE_IP_PORT),
+                )
+                .await;
+        }
+    });
+
+    let info_router_2 = info_router.clone();
```

Comment:
why are you cloning the entire router? this seems wrong I think you want to just have one router and pass it into each closure with an Arc. all of your fields are under a mutex anyways